Detect missing processes

jaki opened the pull request:
Resolves YugaByte/yugabyte-db#1643 and YugaByte/yugabyte-db#1729.

For `yb-ctl`, add various process existence checks:
* Ensure that no master or tserver processes exist at the beginning of
  performing a `create`: added in `ClusterControl.pre_create_checks`
* Ensure that master and tserver processes exist after
  `ClusterControl.start_daemon`s when performing a `create` or `start`:
  addressed by `call_check_for_processes` in
  `ClusterControl.wait_for_cluster`
* Ensure that master and tserver processes continue to exist while
  performing calls to yb-admin in `ClusterControl.wait_for_cluster`:
  addressed by `call_check_for_processes` in `call_check_for_ts`
* Allow PIDs referenced by `postmaster.pid` to refer to already dead
  postgres processes: added in `ClusterControl.stop_process`

jasonyb committed:
Check for existing daemons before create

Among the various sanity checks in `ClusterControl.start_helper`, add
another check to ensure that there are no `yb-` daemons running.  To
facilitate this, have `ClusterControl.for_all_daemons` return the return
values of all of its function calls.

jasonyb committed:
Remodel `start_helper` to `create_sanity_check`

The contents of `start_helper` can be split up into three parts: sanity
checking, preparing data directories, and `start_daemon`s.  Take the
first part, and rename the function appropriately.  Take the other two
parts, and place them in the original sole caller of `start_helper`:
`create_cmd_impl`.

jasonyb committed:
Handle OSError from missing postgres process

During `destroy`, stopping tservers involves looking for
`postmaster.pid` files that contain the PIDs of postgres processes.  In
certain cases, the postgres process may no longer exist.  The desired
behavior is to continue stopping the rest of the tserver and postgres
processes rather than throw an error.  Implement this behavior.

jasonyb committed:
Remove server_counts_map

Things work fine without `server_counts_map`.  `server_counts_map` is
used for calls to `ClusterControl.for_all_daemons`, but passing it into
the call is optional.  Not passing it behaves the same, anyways, so
remove the optional parameter `server_counts_map` from
`ClusterControl.for_all_daemons`.

jasonyb committed:
Add process existence checks to complement RPCs

During the `ClusterControl.wait_for_cluster` step, checks are done using
RPCs to ensure clusters are ready.  In the case that processes go down,
logging could be inaccurate and failure could be delayed.  Add checks
for the existence of processes to address these issues.

jasonyb committed:
Edit some docstrings, log messages, and comments

Expand "\`yb-\`" to "master or tserver".  Simplify and shorten some log
messages.  Unspecify "\`create\`" in a log message that is shared by
`create` and `start`.  Use reStructuredText style return specification
in docstring for `ClusterControl.for_all_daemons`.

rkarthik007 commented:
@WesleyW - would be good to take a look at the fixes here and see if we need to incorporate any in `yugabyte-db`. cc @bmatican @schoudhury

mbautin commented:
@jj-kim please take a look at the test failures at https://travis-ci.org/YugaByte/yugabyte-installation/jobs/556584742

jasonyb committed:
Temporarily set tserver_flags=v=1 in yb-ctl-test.sh

Authored by @mbautin.

jasonyb committed:
Configure loopback IP addresses for Mac and Travis

In the test script, add loopback IP addresses in case of Mac OSX and
TravisCI to cover add node scenarios.  Configuration is from
https://docs.yugabyte.com/latest/quick-start/install/.

jasonyb committed:
Clear create_flags in test script

jasonyb committed:
Unquote create_flags variables in test script

jasonyb committed:
Change a RuntimeError to an ExitWithError

The stack trace is unwanted when erroring for a process going down
during `create` or `start`.  Change the `RuntimeError` to an
`ExitWithError` because it does not display the stack trace.

jasonyb committed:
Clarify create_sanity_check

The name `create_sanity_check` doesn't match what it does quite well.
Rename it to `create_ready_check` and add a docstring clarifying what it
does.

bmatican commented on .gitignore:
```diff
@@ -4,3 +4,5 @@
 *.sw?
 *.bak
 *.rej
+venv/
```
why is there a `venv` here?

jaki replied:
That was @mbautin, commit 963b3d8.  I was using a virtual environment.

bmatican commented on bin/yb-ctl:
```diff
@@ -1246,8 +1246,14 @@ class ClusterControl:
+        try:
+            os.kill(pid, signal.SIGTERM)
+        except OSError as e:
+            if e.errno == 3:
```
is there some os.CONSTANT we can use instead of just 3 :)

bmatican commented on bin/yb-ctl:
```diff
@@ -1297,7 +1303,10 @@ class ClusterControl:
         self.start_daemon(daemon_id)
 
-    def start_helper(self, server_counts_map):
+    def create_ready_check(self):
```
nit: maybe call it `pre_create_checks`

bmatican commented on bin/yb-ctl:
```diff
@@ -1520,18 +1549,19 @@ class ClusterControl:
+        :returns: return values of each function call.
+        :rtype: dict
         """
+        fn_returns = {daemon_type: [] for daemon_type in DAEMON_TYPES}
```
nit: I normally use dict.setdefault(key, []) so I don't have to preinitialize

bmatican commented on test/yb-ctl-test.sh:
```diff
@@ -227,6 +235,15 @@ log "OSTYPE: $OSTYPE"
 log "USER: $USER"
 log "TRAVIS: ${TRAVIS:-undefined}"
 
+if [[ ${TRAVIS:-} == "true" && $OSTYPE == darwin* ]]; then
```
nit: add a comment that mac needs loopback aliases explicitly created

bmatican commented on bin/yb-ctl:
```diff
@@ -1342,6 +1355,18 @@ class ClusterControl:
+            return all(sum(pids.values(), []))
+
+        logging.info("Waiting for master and tserver processes to be up.")
+        if not retry_call_with_timeout(call_check_for_processes, self.options.timeout):
```
instead of the overall timeout here, I'd add a new constant at the top, maybe like 5 seconds, since really, if the processes do not come up in a couple of seconds, they likely crashed

jaki replied:
Forgot to mention that I didn't work on this.  So this timeout is hard-coded and not exposed to the user?

bmatican commented on bin/yb-ctl:
```diff
@@ -1342,6 +1355,18 @@ class ClusterControl:
+            pids = self.for_all_daemons(self.get_pid)
+            if self.args.verbose:
+                logging.info("Daemon PIDs found after starting daemons: {}".format(pids))
+            return all(sum(pids.values(), []))
```
why do we need a sum here? isn't pids already a list, with some items being None? won't all(pids) be enough?

jaki replied:
`pids` is a dict with two keys and list values.  The `sum` flattens it.

jaki replied:
The signature for `for_all_daemons` could change to just return a flattened list, but that would be a loss of information.  One consequence is that the log message above would be less clear.

bmatican replied:
ah makes sense, my bad, didn't realize the signature

bmatican commented on bin/yb-ctl:
```diff
@@ -1357,6 +1382,9 @@ class ClusterControl:
 
         def call_check_for_ts(should_get_error):
+            if not call_check_for_processes(should_get_error):
+                raise ExitWithError("At least one master or tserver process is down.")
```
I think we should also have one more debug function where, if we know some processes are down, we log out the respective yb-master.err / yb-tserver.err file for said process, in the hopes of highlighting why it crashed. We can then use that function both here and in line 1367, since at the end of that loop, we know we should be failing as well

bmatican commented on bin/yb-ctl:
```diff
@@ -1520,18 +1549,19 @@ class ClusterControl:
-            else:
-                num_servers = self.get_number_of_servers(daemon_type)
-
+            num_servers = self.get_number_of_servers(daemon_type)
```
wait, how does this work during `create`?? we won't have the respective directories created per node, for `get_number_of_servers` to actually be able to return the correct numbers...

jaki replied:
It goes into the first if case, which uses replication factor.

bmatican replied:
oooh, because of `        self.creating_cluster = True`!

jasonyb committed:
Address some of @bmatican's requested changes

* Use constant `os.errno.ESRCH` rather than literal `3`
* Rename `create_ready_check` to `pre_create_checks`
* Use `dict.setdefault` in `for_all_daemons`
* Add Mac needs loopback aliases comment.

jasonyb committed:
Log out error file of missing processes

When failing from processes not coming up or unexpectedly going down,
dump the error logs of the missing processes to stderr.

jasonyb committed:
Take some logging out of verbose

Remove the `verbose` condition from some PID log reporting so that the
PID logging always gets reported upon an error.

jasonyb committed:
Reduce timeout on waiting for processes to start

Waiting for processes to start uses the same timeout as waiting for good
RPC calls.  Rather, waiting for processes should have a lower timeout,
so make a separate, smaller constant value for this timeout.

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
 
+    def dump_missing_process_error_logs(self):
+        pids = self.for_all_daemons(self.get_pid)
+        for daemon_type, pid_list in pids.items():
```
nit: prefer iteritems over items

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+    def dump_missing_process_error_logs(self):
+        pids = self.for_all_daemons(self.get_pid)
+        for daemon_type, pid_list in pids.items():
+            for index, pid in enumerate(pid_list, 1):
```
why do we start at 1?

jaki replied:
Daemon indices start at 1.

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+                    node_base_dirs = self.get_base_node_dirs(daemon_id.index)
+                    first_base_dir = node_base_dirs[0]
+                    log_name = daemon_id.daemon_type + ".err"
+                    log_path = os.path.join(first_base_dir, log_name)
```
we should already be constructing this path when we start the daemon, can we just make this be a function to pick up the .err / .out path?

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+                    first_base_dir = node_base_dirs[0]
+                    log_name = daemon_id.daemon_type + ".err"
+                    log_path = os.path.join(first_base_dir, log_name)
+                    if os.path.exists(log_path):
```
maybe for missing .err file, we should also log something like `process failed without writing out any error logs..`

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+                    if os.path.exists(log_path):
+                        with open(log_path) as log:
+                            print("Viewing log at {}:".format(log_path), file=sys.stderr)
+                            if sys.version_info[0] >= 3:
```
we also have a spot where we dump the log to stderr if we have errors, can we make a shared function for both of these

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+            return all(sum(pids.values(), []))
+
+        logging.info("Waiting for master and tserver processes to be up.")
+        if not retry_call_with_timeout(call_check_for_processes, MAX_PROCESS_WAIT_SECONDS):
```
@mbautin you ok with this particular wait for processes to come up to not be tied to the overall `--timeout` option? This is how much time to wait for the cluster processes to all be up, which we'd expect to happen pretty fast or not at all...

I think we default that to 45s, probably due to previous `initdb` slowdown. I had suggested to @jj-kim to make this a separate constant with lower value (5-10s).

bmatican commented on bin/yb-ctl:
```diff
@@ -1335,13 +1348,43 @@ class ClusterControl:
+        # Wait for master and tserver processes to be ready.
+        def call_check_for_processes(should_get_error):
+            pids = self.for_all_daemons(self.get_pid)
+            logging.info("PIDs found: {}".format(pids))
```
nit: this will be logged all per check, which might be spammy...why not just log it in the error checking if outside? or even better, just in `dump_missing_process_error_logs`?

bmatican reviewed (changes requested):
Mostly minor stuff, should be good to go after this, I think.

Commit:
Improve logging for wait_for_cluster

* Move "PIDs found" logging out of the main code path to the respective
  error code paths
* Change `dict.items` call to `dict.iteritems` for lazy loading
* Add logging for when a process is found down but there is no error log
* Change the log message for waiting for processes to "be up" to "come
  up" in order to match the "come up" of a later log message.

Commit:
Create two methods to unduplicate code

There is some code duplication in `ClusterControl`.  Create methods to
address this: `get_log_path` and `keep_and_dump_log`.  The only other
change should be an extra "Viewing log at" message for the main log
`self.log_file`.

bmatican commented on bin/yb-ctl:
```diff
@@ -1348,23 +1365,33 @@ class ClusterControl:
         if not retry_call_with_timeout(call_change_config, self.options.timeout):
             raise ExitWithError("Could not change master config...")
 
+    def keep_and_dump_log(self, log_path):
```
nit: can we rename this to just `dump_file_to_stderr`, since having a `keep_` prefix makes me think we would also have an API that removes it.

bmatican reviewed (approved):
Let's check with @mbautin real quick offline, if he thinks this new timeout being shorter and not CLI changeable is fine with him to.

Commit:
Rename keep_and_dump_log to dump_file_to_stderr

The method `keep_and_dump_log` is actually more generic than the name
implies.  Rename `log` related names to `file` related names.  Also
update the messages talking about a file "at" some path to not use "at"
because that is more applicable when path is a directory.

Commit:
Add cluster up timeout as cmdline option

The timeout for waiting for processes of the cluster to come up is
hard-coded and not exposed to the user.  Expose it to the user as an
`argparse` option, just like the existing timeout option.  Rename the
timeout variables appropriately.

mbautin commented on bin/yb-ctl:
```diff
@@ -540,7 +541,8 @@ class ClusterOptions:
         self.max_daemon_index = 20
         self.num_shards_per_tserver = None
-        self.timeout = None
+        self.timeout_yb_admin = None
```
`timeout_yb_admin_sec`

mbautin commented on bin/yb-ctl:
```diff
@@ -106,7 +106,8 @@ YEDIS_DEFAULT_PORT = 6379
 SLEEP_TIME_IN_SEC = 1
-MAX_WAIT_SECONDS = 45
+MAX_YB_ADMIN_WAIT_SECONDS = 45
+MAX_CLUSTER_UP_WAIT_SECONDS = 10
```
Instead of "cluster up", this should say "processes running", because that's what we are testing.
E.g. MAX_WAIT_FOR_PROCESSES_RUNNING_SEC

mbautin commented on bin/yb-ctl:
```diff
@@ -540,7 +541,8 @@ class ClusterOptions:
         self.num_shards_per_tserver = None
-        self.timeout = None
+        self.timeout_yb_admin = None
+        self.timeout_cluster_up = None
```
`timeout_processes_running_sec`

mbautin commented on bin/yb-ctl:
```diff
@@ -906,8 +927,11 @@ class ClusterControl:
+            "--timeout-yb-admin", default=MAX_YB_ADMIN_WAIT_SECONDS, type=float,
+            help="Timeout in seconds for operations that call yb-admin and wait on the cluster")
+        self.parser.add_argument(
+            "--timeout-cluster-up", default=MAX_CLUSTER_UP_WAIT_SECONDS, type=float,
```
timeout-processes-running-sec

mbautin commented on bin/yb-ctl:
```diff
@@ -906,8 +927,11 @@ class ClusterControl:
         self.parser.add_argument(
-            "--timeout", default=MAX_WAIT_SECONDS, type=float,
-            help="Timeout in seconds for operations that wait on the cluster")
+            "--timeout-yb-admin", default=MAX_YB_ADMIN_WAIT_SECONDS, type=float,
```
timeout-yb-admin-sec

Commit:
Rename timeout variables

Rename the timeout variables to be more clear, as suggested by @mbautin.
